File: python/00-parciales/01-pensiones/main.py
import random

# Ejercicio 2: encontrar TAD adecuado
# - simular durante 5 horas
# - dos cajeros
# - tiempo de atención: 6 y 4 minutos
# - frecuencia de llegada de pensionados: 3 minutos


class Nodo:
    def __init__(self, item: int, sig=None) -> None:
        self.__item = item
        self.__sig = sig

    def get_item(self):
        return self.__item

    def get_siguiente(self):
        return self.__sig

    def set_siguiente(self, sig: "Nodo"):
        self.__sig = sig


class Cola:
    __cab: Nodo | None
    __cola: Nodo | None
    __long: int

    def __init__(self) -> None:
        self.__cab = None
        self.__cola = None
        self.__long = 0

    def insertar(self, item):
        nuevo_nodo = Nodo(item)

        if self.vacia():
            self.__cab = nuevo_nodo
        else:
            self.__cola.set_siguiente(nuevo_nodo)
        self.__cola = nuevo_nodo
        self.__long += 1

    def eliminar(self):
        if self.vacia():
            return None
        temp = self.__cab
        self.__cab = temp.get_siguiente()
        self.__long -= 1
        return temp.get_item()

    def recorrer(self):
        actual = self.__cab
        while actual:
            print(actual.get_item(), end=" ")
            actual = actual.get_siguiente()

    def vacia(self):
        return self.__long == 0


if __name__ == "__main__":
    cajeros = [Cola(), Cola()]
    tiempo_atencion = [6, 4]
    antendidos = [0, 0]
    tiempo_llegada = 3
    n = 2

    for reloj in range(300):
        if reloj % 3 == 0:
            cajeros[random.randint(0, n - 1)].insertar(1)
        i = 0
        while i < n:
            if reloj % tiempo_atencion[i] and not cajeros[i].vacia():
                cajeros[i].eliminar()
                antendidos[i] += 1
            i += 1

    print(f"Cajero 1 atendio: {antendidos[0]} pensionados.")
    print(f"Cajero 2 atendio: {antendidos[1]} pensionados.")

File: cpp/00-parciales/01-pensiones/main.cpp
#include <cstdlib>
#include <ctime>
#include <iostream>

class Nodo {
private:
  int _item;
  Nodo *_sig;

public:
  Nodo(int item, Nodo *sig = nullptr) : _item(item), _sig(sig) {}
  int get_item() { return _item; }
  Nodo *get_sig() { return _sig; }
  void set_sig(Nodo *sig) { _sig = sig; }
};

class Cola {
private:
  Nodo *_cab;
  Nodo *_cola;
  int _long;

public:
  Cola() {
    _cab = nullptr;
    _cola = nullptr;
    _long = 0;
  }
  ~Cola() {
    while (!vacia()) {
      eliminar();
    }
  }
  void insertar(int item) {
    Nodo *nuevo_nodo = new Nodo(item);
    if (vacia()) {
      _cab = nuevo_nodo;
    } else {
      _cola->set_sig(nuevo_nodo);
    }
    _cola = nuevo_nodo;
    _long++;
  }
  int eliminar() {
    auto temp = _cab;
    _cab = _cab->get_sig();
    _long--;
    int temp_item = temp->get_item();
    delete temp;
    return temp_item;
  }
  void recorrer() {
    auto actual = _cab;
    while (actual) {
      std::cout << actual->get_item() << " ";
      actual = actual->get_sig();
    }
    std::cout << std::endl;
  }
  bool vacia() { return _long == 0; }
};

int numero_random(int ref) {
  std::srand(std::time(nullptr));
  return std::rand() % ref;
}

int main() {
  const int N = 2;
  auto cajeros = new Cola *[2]{new Cola(), new Cola()};
  auto tiempo_atención = new int[2]{6, 4};
  auto atendidos = new int[2]{0, 0};
  int duracion = 60 * 5;
  int reloj = 0;

  while (reloj < duracion + 1) {
    if (reloj % 3 == 0 and numero_random(N) == 1) {
      cajeros[numero_random(N)]->insertar(1);
    }
    int i = 0;
    while (i < N) {
      if (reloj % tiempo_atención[i] == 0 and numero_random(N) == 1) {
        cajeros[i]->eliminar();
        atendidos[i]++;
      }
      i++;
    }
    reloj++;
  }

  std::cout << "El cajero 1 atendió: " << atendidos[0] << " pensionados"
            << std::endl;
  std::cout << "El cajero 2 atendió: " << atendidos[1] << " pensionados"
            << std::endl;
  return 0;
}
